### PR TITLE
test: add e2e test cases for server-info plugin

### PR DESCRIPTION
**Pull request by fhuzero**
This PR added some e2e test cases to cover the usage of `server-info` plugin. A new k8s tunnel `apisixControlTunnel` is added to support the retrieval of server info from `/v1/server_info` through Control API.

### CODE REVIEW

**Comment by tokers** [comment]
@fhuzero Hi! Does this PR ready to review?

**Comment by fhuzero** [comment]
> @fhuzero Hi! Does this PR ready to review?

Sad that the code failed to pass some ci checks so probably it's not ready to be merged. When I run test cases in Minikube on my local machine, though the test for retrieving server info succeed most times, it fails randomly because of some errors such as `waiting for apisix ready`, `no etcd`, `TLS handshake timeout`, etc. These errors will recover later without intervention. It would be great if someone can give me guidance on how to avoid the random failure.

**Comment by tokers** [comment]
> > @fhuzero Hi! Does this PR ready to review?
> 
> Sad that the code failed to pass some ci checks so probably it's not ready to be merged. When I run test cases in Minikube on my local machine, though the test for retrieving server info succeed most times, it fails randomly because of some errors such as `waiting for apisix ready`, `no etcd`, `TLS handshake timeout`, etc. These errors will recover later without intervention. It would be great if someone can give me guidance on how to avoid the random failure.

We recommend you to use [Kind](https://kind.sigs.k8s.io/) as the Cluster to run e2e test suites, also, you may need to push these images to minikube before you start to run them, too long time for pulling images might lead to some timeout assertions failed.

**Comment by fhuzero** [comment]
> > > @fhuzero Hi! Does this PR ready to review?
> > 
> > 
> > Sad that the code failed to pass some ci checks so probably it's not ready to be merged. When I run test cases in Minikube on my local machine, though the test for retrieving server info succeed most times, it fails randomly because of some errors such as `waiting for apisix ready`, `no etcd`, `TLS handshake timeout`, etc. These errors will recover later without intervention. It would be great if someone can give me guidance on how to avoid the random failure.
> 
> We recommend you to use [Kind](https://kind.sigs.k8s.io/) as the Cluster to run e2e test suites, also, you may need to push these images to minikube before you start to run them, too long time for pulling images might lead to some timeout assertions failed.

Thanks for your recommendation! I did follow three steps listed in [Quick Start](https://github.com/apache/apisix-ingress-controller/tree/master/test/e2e#quick-start) in Apisix ingress controller e2e test suites Readme file. When running test for the first time, I created a cluster by Kind and all test cases passed. After that, I used `E2E_SKIP_BUILD=1` to skip the first two steps. Then the aforementioned errors occurred randomly. Please let me know if I overlooked anything.

File: test/e2e/plugins/server-info.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+		Kubeconfig:              scaffold.GetKubeconfig(),
+		APISIXConfigPath:        "testdata/apisix-gw-config.yaml",
+		APISIXDefaultConfigPath: "testdata/apisix-gw-config-default.yaml",
+		IngressAPISIXReplicas:   1,
+		HTTPBinServicePort:      80,
+		APISIXRouteVersion:      "apisix.apache.org/v2alpha1",
+	}
+	s := scaffold.NewScaffold(opts)
+	ginkgo.It("check server info", func() {
+		err := setServerInfoPluginStatus(opts.APISIXConfigPath, true)
```
Comment (tokers) [inline test/e2e/plugins/server-info.go:41]: Timing is too late, apisix deployment was created already before the `It` block running.

**Comment by tokers** [comment]
@fhuzero Maybe you can simply this PR, as we always enable `server-info` plugin, we may just enable it and check its output.

**Comment by tokers** [comment]
@fhuzero Also, please resolve the conflicts.

**Comment by fhuzero** [comment]
> @fhuzero Maybe you can simply this PR, as we always enable `server-info` plugin, we may just enable it and check its output.

Cool. Actually I have been doubting the need of disabling the plugin and spent a lot of time dealing with the case when the plugin is disabled. Seems that I'd better raise this question and discuss it earlier :P

**Comment by tokers** [comment]
> > @fhuzero Maybe you can simply this PR, as we always enable `server-info` plugin, we may just enable it and check its output.
> 
> Cool. Actually I have been doubting the need of disabling the plugin and spent a lot of time dealing with the case when the plugin is disabled. Seems that I'd better raise this question and discuss it earlier :P

Just take your time :)

File: test/e2e/plugins/server-info.go [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+		HTTPBinServicePort:      80,
+		APISIXRouteVersion:      "apisix.apache.org/v2alpha1",
+	}
+	s := scaffold.NewScaffold(opts)
+
+	ginkgo.It("enable server-info plugin", func() {
+		serverInfoKey := [...]string{"etcd_version", "up_time", "last_report_time", "id", "hostname", "version", "boot_time"}
+		serverInfo, err := s.GetServerInfo()
+		assert.Nil(ginkgo.GinkgoT(), err)
+		if assert.NotNil(ginkgo.GinkgoT(), serverInfo) {
```
Comment (tokers) [inline test/e2e/plugins/server-info.go:39]: Actually `serverInfo` must not empty. So here should not use such a conditional check.

File: test/e2e/scaffold/k8s.go [hunk 3]
```diff
@@ -184,6 +184,30 @@ func (s *Scaffold) EnsureNumApisixUpstreamsCreated(desired int) error {
...
 }
 
+// GetServerInfo collect server info from "/v1/server_info" (Control API) exposed by server-info plugin
+func (s *Scaffold) GetServerInfo() (map[string]interface{}, error) {
+	u := url.URL{
+		Scheme: "http",
+		Host:   s.apisixControlTunnel.Endpoint(),
+		Path:   "/v1/server_info",
+	}
+	resp, err := http.Get(u.String())
```
Comment (tokers) [inline test/e2e/scaffold/k8s.go:194]: deferred to close the body.

Comment (fhuzero) [reply test/e2e/scaffold/k8s.go:194]: I added a `defer` statement if no error is returned.

**Review by tokers — APPROVED** [review]

**Pull request merged**